### PR TITLE
fix(spec): an embedded struct contributes its fields

### LINKED ISSUE #487: An embedded struct contributes no fields to the schema
## What

A struct that embeds another documents **none** of the embedded fields — neither the promoted ones nor a tagged embed's nested object.

```go
type Base struct {
	ID   int
	Kind string `json:"kind"`
}

type Named struct {
	Label string
}

type Item struct {
	Base                 // promoted: ID and kind are Item's own JSON fields
	Named  `json:"named"` // tagged: nested under "named"
	Extra  string
}
```

What `encoding/json` actually produces:

```json
{"ID":0,"kind":"","named":{"Label":""},"Extra":""}
```

What apispec produces:

```yaml
Item:
  type: object
  properties:
    Extra: {type: string}
```

Three of the four wire fields are missing. A client generated from this document cannot read `ID`, `kind` or `named` at all, and there is nothing in the output to say they were dropped.

## Scope

Embedding is not a corner case — it is how Go composes response types, and a shared `Base`/`Meta`/`Envelope` embed is one of the most common shapes in a real API. Any response whose type embeds anything is under-documented by however much the embed carries.

## What is already correct (checked, so it is not re-investigated)

Default field NAMING with no `json` tag is right, and matches the encoder exactly:

| field | Go emits | apispec |
|---|---|---|
| `Tagged string \`json:"tagged_name"\`` | `tagged_name` | `tagged_name` ✓ |
| `Untagged string` | `Untagged` | `Untagged` ✓ |
| `ID int` | `ID` | `ID` ✓ |
| `HTTPCode int` | `HTTPCode` | `HTTPCode` ✓ |
| `unexp string` | omitted | omitted ✓ |
| `Skipped string \`json:"-"\`` | omitted | omitted ✓ |
| `OmitOnly string \`json:",omitempty"\`` | `OmitOnly` | `OmitOnly` ✓ |

Verified against `json.Marshal` on the same struct. So the field-name rule needs no work; it is only embedding that is lost.

## Sub-cases worth covering in the fix

1. **Plain embed** → fields promoted into the parent's properties, with the embedded type's own tags applied (`kind`, not `Kind`).
2. **Tagged embed** (`Named \`json:"named"\``) → a nested object under that name, NOT promoted.
3. **Embedded pointer** (`*Base`) → same promotion as the value form.
4. **Name collision** — a promoted field whose name matches one the outer struct declares: Go's rule is that the shallower depth wins and equal depth cancels both out.
5. **Embedded non-struct** (`type ID string` embedded) → contributes a field named for the type.
6. **Required/`allOf`** — whether promotion is emitted as flattened properties or as `allOf: [$ref: Base, {...}]`. Flattening matches the wire; `allOf` keeps the component reusable. Worth deciding deliberately, since `$ref`-ing the embed is what most hand-written specs do.

## Acceptance

- A `testdata/embedded_structs` fixture covering at least 1–4, with a structural test asserting the property set matches what `encoding/json` emits for the same type.
- No change to the non-embedded naming behaviour above.

## Notes

`metadata.Type` already records `Embeds []int` (the embedded type names), so the facts are present — this is a mapper gap, not a metadata one. Raised while working on #485.

### PR DESCRIPTION
**Pull request by ehabterra**
Closes #487.

## The gap

A struct that embeds another documented **none** of what the embed carries — the mapper walked `Type.Fields`, and an embed is recorded in `Type.Embeds`.

```go
type Item struct { Base; Named `json:"named"`; *Meta; Ref; Extra string }
```
```
encoding/json: {"ID":0,"kind":"","named":{"Label":""},"Page":0,"Ref":"","Extra":""}
apispec:       {"Extra": …}
```

Embedding is how Go composes response types, so any response embedding a shared `Base`, `Meta` or `Envelope` was under-documented by exactly the embed's contents, silently.

## Rules implemented — all measured against the encoder

Every expectation in the tests is what `json.Marshal` actually emits for the same type, captured by **running it**:

| shape | behaviour |
|---|---|
| untagged embed | promotes the fields, carrying their own tags (`Kind string \`json:"kind"\`` → `kind`) |
| tagged embed | does **not** promote — an ordinary field named by the tag, holding the object |
| embedded pointer | promotes exactly as a value does |
| embedded non-struct | one field named for the type |
| `json:"-"` embed | dropped whole |
| name collision | shallowest wins; at equal depth exactly one **tagged** field wins, else Go sends neither and neither do we (golden rule #7) |

## One metadata fact was required

`Type.EmbedTags`. Without the embed's tag, a tagged embed is indistinguishable from a plain one — and promoting it would publish the embedded fields at top level under names the wire never uses, turning a *missing* answer into a *wrong* one. A parallel slice keeps every existing reader of `Embeds` working, and `omitempty` keeps it out of the goldens of types that embed nothing tagged.

## Two regressions of my own, caught by the fixture A/B

Worth stating because neither was predictable from reasoning:

1. **An embed we cannot see through must contribute nothing.** `encoding/json` promotes an external type's exported fields; we can't see them. Claiming a field named for the type published a `Userinfo` property for `url.Userinfo` — which has **no exported fields at all**.
2. **The blank marker `_ struct{} \`validate:"…"\`** carries struct-level validation (#166). It never serialises, so the effective-field set correctly excludes it — and the note therefore vanished. It now comes from the type's own fields, outside the loop.

## Measurements

| | |
|---|---|
| fixtures | **123 of 123 byte-identical** |
| gitea | **+27 properties across 5 schemas, −0** |
| 280-path service | **+117 properties across 4 schemas, −0** |
| 2 smaller projects | byte-identical |

Verified in gitea's own source: `structs.Commit` embeds `*CommitMeta` (gains `sha`, `url`, `created`); `structs.CommitUser` embeds `Identity` (gains `name`, `email`). On the 280-path service every changed schema is a `*Detail` type embedding its base.

## Two things stated rather than hidden

**Ordering.** Effective fields come out fields-then-embeds, which is not always Go's declaration order. Metadata keeps `Fields` and `Embeds` as two lists with no record of how they interleaved, so true declaration order needs another fact. It reaches the document only through `required`, whose order carries no meaning — which is why every fixture is byte-identical.

**`maxEmbedDepth` stays a constant**, and I measured before deciding: on gitea (~232 anonymous embedded fields) the walk reaches depth **1** and never 2, so six is six times the deepest observed use, and the cycle guard is the visited set rather than the depth. The limits that *are* configurable became so because a real project measurably needed a different number; this one would be a knob with nothing on the other end. Happy to add it if you'd rather — it's a flag, an engine field and a UI control.

Suite, `make lint`, coverage ratchet (96.0%) pass.

## Summary by CodeRabbit

* **Bug Fixes**
  * OpenAPI schemas now correctly represent fields from embedded Go structs, including promoted, nested, pointer-based, and tagged fields.
  * Embedded fields now follow `encoding/json` behavior for field naming, shadowing, visibility, and ignored fields.
  * Improved handling prevents incorrect or missing fields in generated schemas.
* **Tests**
  * Added comprehensive coverage for embedded-struct schema generation and field-resolution edge cases.

### CODE REVIEW

File: internal/spec/embedded_fields.go [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+			// show it is not a struct; otherwise the branch above applies.
+			*out = append(*out, effectiveField{
+				field:  metadata.Field{Name: meta.StringPool.Get(embeddedFieldName(name)), Type: embedIdx},
+				owner:  typ,
+				depth:  depth,
+				tagged: false,
+			})
+			continue
+		}
+		collectJSONFields(meta, embedded, depth+1, visited, out)
```
[the remark this answers is not in the document]
Comment (ehabterra) [reply internal/spec/embedded_fields.go]: Confirmed against the encoder before fixing, and it is exactly as described — fixed in aef45f3.

```
go:      {"Extra":""}
apispec: [ID Extra]
```

`Outer{A; B}` where both embed `Base`: one visited set for the whole walk saw `Base` once, so `ID` had a single claimant and survived collision resolution. encoding/json sees **two** untagged claimants at equal depth and sends neither — so the schema documented a property the encoder never emits, which is the failure mode this whole area is about.

The set is now cloned per path, so two routes to one type stay two candidates and the existing collision rule does the rest. Regression test added with the `{"Extra":""}` expectation captured by running `json.Marshal`.

File: internal/spec/embedded_fields.go [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+// reaches depth 1 and never 2, so six is already six times the deepest observed
+// use. The cycle guard is the visited set below, not this; Go forbids recursive
+// embedding outright, so reaching either would mean metadata that cannot
+// describe a program that compiles.
+//
+// The limits that ARE configurable became so because a real project measurably
+// needed a different number (MaxNodesPerTree, MaxInstancesPerKey). Adding a
+// flag, an engine field and a UI control for this one would be a knob with
+// nothing on the other end.
+const maxEmbedDepth = 6
```
[the remark this answers is not in the document]
Comment (ehabterra) [reply internal/spec/embedded_fields.go]: Agreed, and removed entirely in aef45f3 rather than raised.

You are right that it is a correctness limit, not a safety one: seven untagged embeds is legal Go whose leaf fields encoding/json promotes, and the cap would have dropped them **silently**. Verified an eight-level chain now resolves to `[Deep]`, matching the encoder.

The per-path visited set is the real guarantee, and it terminates without a cap on its own — each step adds a type to that path's set and never revisits it, over a finite set of declared types. Go forbids recursive embedding outright, so the guard only matters for metadata that could not describe a compiling program.

Worth noting this also settles a question that was open on my side: the repo owner had asked whether the cap should be **configurable**. With the per-path guard doing the work there is nothing left to configure, which is a better answer than a knob — my earlier measurement (gitea reaches depth 1, never 2) said a knob would have nothing on the other end, and this removes the limitation as well.

File: internal/metadata/metadata.go [hunk 3]
```diff
@@ -1095,8 +1095,10 @@ func processStructFields(structType *ast.StructType, pkgName string, metadata *M
 		}
 
 		if len(field.Names) == 0 {
-			// Embedded (anonymous) field
+			// Embedded (anonymous) field. The tag travels with it because it is
+			// what tells promotion from nesting — see Type.EmbedTags.
 			t.Embeds = append(t.Embeds, metadata.StringPool.Get(fieldType))
+			t.EmbedTags = append(t.EmbedTags, metadata.StringPool.Get(tag))
```
[the remark this answers is not in the document]
Comment (ehabterra) [reply internal/metadata/metadata.go:1118]: Correct, and verified: `StringPool.Get("")` returns `NoString` (-1), so an untagged embed contributes `-1` and the slice is non-empty — its own `omitempty` never fires, and every type that embeds anything would write `embed_tags: [-1, -1]` into its metadata.

Fixed in aef45f3 by filling it positionally as before and then dropping it whole when **no** embed carried a tag, which is nearly always. Mixed cases keep their alignment, and `embedded_fields.go` already reads a missing entry as an empty tag (the `i < len(typ.EmbedTags)` guard), so the nil case needs nothing.

`EmbedAt` is deliberately kept in every case: unlike a tag, a position is meaningful for every embed, and it is what makes declaration order recoverable.
